### PR TITLE
disable pause and rewinds for Intel HDaudio platforms

### PR DESCRIPTION
**Pull request by plbossart**
Tentative RFC to combine the topics @ranj063 and I discussed to make a better use of hardware and hopefully reduce power consumption.

Disabling the rewinds was suggested back in 2015..2018, see following links for initial patches:

https://lore.kernel.org/alsa-devel/1521561668-28613-1-git-send-email-sriramx.periyasamy@intel.com/#r
https://lore.kernel.org/alsa-devel/1494896518-23399-1-git-send-email-subhransu.s.prusty@intel.com/

The main difference is that since there was no agreement to add a new hw_params flags these capabilities can only enabled via Kconfig or kernel parameters. If someone uses PulseAudio with rewinds disabled it's not fatal but the user interaction will suck, e.g. with delayed volume changes and notifications. No side effects are expected with CRAS, JACK and PipeWire.

### CODE REVIEW

File: sound/soc/sof/intel/hda-pcm.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-pcm.c]: why not check direction, substream and add the check? theres no rewind for capture anyway right?

File: sound/soc/sof/intel/Kconfig
Comment (ranj063) [inline sound/soc/sof/intel/Kconfig]: can we use DMI_L1 directly instead of adding another kconfig? if ALWAYS_DMI_L1 is enabled, we should automatically disable pause no?

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: I don't know, this is why it's tagged as FIXME :-)

File: sound/soc/sof/intel/Kconfig
Comment (plbossart) [reply sound/soc/sof/intel/Kconfig]: I did it the other way around.

a) you may want to disable PAUSE if you feel like it for tests.
b) you may want to explore what happens when DMI L1 is enabled, even without pause
c) if you disable pause, the code automatically selects DMI L1 since it's safe to do so.

**Comment by plbossart** [comment]
Minor changes: I made the pause disable a generic feature for all SOF drivers. A distribution should not have to select if pause works for ACPI and not for PCI.

ALSA level tests work with SPIB, they completely fails with Fedora34/PipeWire. There are no traces of any .ack and the trace messages keep on flooding the console. Something's missing in the handling of the hw_sync_ptr.

Edit: same results with Ubuntu and Pulseaudio, playing test streams with the gnome settings app generates a flood of TRACE_MSG.

````
options snd-sof-intel-hda-common disable_rewinds=1
````

````
[  559.927254] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  559.943235] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  559.943268] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  559.959232] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  559.959263] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  559.975231] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
````

File: sound/soc/sof/intel/hda-pcm.c
Comment (lyakh) [inline sound/soc/sof/intel/hda-pcm.c:295]: doesn't this contradict the commit description?

File: sound/soc/sof/intel/hda-pcm.c
Comment (lyakh) [inline sound/soc/sof/intel/hda-pcm.c]: `if (!spib) spib = buf_size;` would be enough

File: sound/soc/sof/intel/hda-pcm.c
Comment (lyakh) [inline sound/soc/sof/intel/hda-pcm.c:35]: is there any specific reason to have this parameter (and another one in an earlier patch) both in Kconfig and as a module parameter? Wouldn't just a module parameter be enough?

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:295]: Gah, yes it's inverted. Nice catch.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: we have to revisit this formula, every time I see it I think it's wrong then convince myself it's correct.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:35]: this is what we've done for all parameters. The Kconfig gives you a default, but you can override it at run time. standard in all ALSA HDAudio drivers.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:295]: fixed

File: sound/soc/sof/intel/hda-pcm.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-pcm.c]: thisis a duplicate of line 170

**Comment by ranj063** [comment]
@plbossart the patches look OK but what would not be clear for everyone would be that there are 3 kernel configs to mess with now and the inter-dependencies between them.
1. ALWAYS_DMI_L1
2. DISABLE_REWINDS
3. DISABLE_PAUSE

Its simple to tell that if all 3 are not chosen, we fallback to current behaviour. 
For example in terms of power 1 & 3 should be the same. whereas 2 + 1 or 2 +3 would give extra savings.

**Comment by plbossart** [comment]
> @plbossart the patches look OK but what would not be clear for everyone would be that there are 3 kernel configs to mess with now and the inter-dependencies between them.
> 
> 1. ALWAYS_DMI_L1
> 2. DISABLE_REWINDS
> 3. DISABLE_PAUSE
> 
> Its simple to tell that if all 3 are not chosen, we fallback to current behaviour.
> For example in terms of power 1 & 3 should be the same. whereas 2 + 1 or 2 +3 would give extra savings.

I was actually tempted to remove the ALWAYS_DMI_L1 as a Kconfig. It's dangerous and far from explicit.

That would leave us with two cases that have a separate impact on power.

**Comment by ranj063** [comment]
> I was actually tempted to remove the ALWAYS_DMI_L1 as a Kconfig. It's dangerous and far from explicit.
> 
> That would leave us with two cases that have a separate impact on power.

Yes, I vote for it. It cannot be chosen without disabling pause anyway!

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: fixed

**Comment by plbossart** [comment]
> > I was actually tempted to remove the ALWAYS_DMI_L1 as a Kconfig. It's dangerous and far from explicit.
> > That would leave us with two cases that have a separate impact on power.
> 
> Yes, I vote for it. It cannot be chosen without disabling pause anyway!

done.

**Comment by plbossart** [comment]
This is starting to take shape with good results w/ aplay, aplay -M (mmap), PulseAudio and PipeWire.

We still need to have power consumption results from @ranj063 and @harshapriya-n.

File: sound/soc/sof/intel/hda-pcm.c
Comment (lyakh) [reply sound/soc/sof/intel/hda-pcm.c:35]: Ok, but apart from being a common approach, who builds custom kernels these days? And distros will have this parameter set one way or another for all their kernels anyway. Much easier to change under modprobe.d

Comment (lyakh) [reply sound/soc/sof/intel/hda-pcm.c:35]: but in fact, in this specific case, you write, that it (also) depends on the audio user-space (PA vs. PipeWire vs...) and it's exactly distros who will configure that default user-space. But if the user opts for the non-default option? E.g. if the distro's default is PipeWire so they enable this feature in Kconfig. But then if the user wants to install and use PulseAudio instead, they'll have to manually add the kernel parameter? Or the distro will have to know to do that during PA installation... Cannot this be done automatically?

**Review by lyakh — APPROVED** [review]

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:35]: It's not realistic to assume that kernel configurations will change if the user decides to revert to PulseAudio. Either the user will have to edit /etc/modprobe.d/, or the script to switch from PipeWire to PulseAudio does it. see https://fedoraproject.org/wiki/Changes/DefaultPipeWire

Or they don't use the parameter to force rewinds to be enabled. In practice I don't see the difference.

File: sound/soc/sof/intel/hda-pcm.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda-pcm.c:298]: Hmm, this doesn't seem right. Whether DMI-L1 is always enabled shouldn't be related to the question on whether a astream is L1 (aka D0i3x) compatible or not.

File: sound/soc/sof/Kconfig
Comment (kv2019i) [inline sound/soc/sof/Kconfig:55]: This "On some platforms this will enable power saving" is a bit vague. Can we add a bit more text on how this relates to power savings. Or just mention the "simplify transitions" bit, that is easier to understand.

File: sound/soc/sof/intel/hda-pcm.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda-pcm.c]: Some debug code leftovers here...

**Review by kv2019i — CHANGES_REQUESTED** [review]
Wow, quite a patchset, DMI changes, addtion to ALSA uapi, new HDA programming, et al all in one PR! :) 

Looks pretty good and I guess this code was tested in the earlier iteration already. I did find a few minor things that do need to be fixed before merge. I'll want to also test this a bit more. 

It's a bit iffy we need to have so many options (we are already very Kconfig-rich ), but yeah, I guess there's no nice way around than giving up on the power savings or without cutting features altogether. For pause, I wonder if it would be easier to just drop it. Not aware of any real-life apps that actually use it.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: gah yes. thanks for spotting this.

File: sound/soc/sof/Kconfig
Comment (plbossart) [reply sound/soc/sof/Kconfig:55]: will reword.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: fixed

File: sound/soc/sof/Kconfig
Comment (plbossart) [reply sound/soc/sof/Kconfig:55]: please double check the updated text.

**Comment by plbossart** [comment]
> For pause, I wonder if it would be easier to just drop it. Not aware of any real-life apps that actually use it.

I suggested such a change earlier. There was no consensus, and it turns out that Windows does support pause so the firmware will have to be built with pause support.

Also it's been my experience that pause does bring a bit of spice to the usual sequences, e.g. when you do a pause_release after a system suspend, you end-up with different combinations of prepare/trigger. It's useful IMHO to make sure those sequences are supported for completeness. Put differently, the code will look better if we support pause, it's a great pipe cleaner and a very good way to find bugs. I see a lot of value in keeping it alive for development and CI/daily stress tests.

File: sound/soc/sof/intel/hda-pcm.c
Comment (kv2019i) [reply sound/soc/sof/intel/hda-pcm.c:298]: @plbossart  ping?

**Review by kv2019i — CHANGES_REQUESTED** [review]
I was looking for response to one comment in the DMI-L1 patches in the series. Otherwise looking good.

File: sound/soc/sof/intel/hda-pcm.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:298]: my comment seems to have been lost?
I don't get the point, this adds a filter that if true don't require the rest to be evaluated. If you always enable DMI L1, you don't need to know the direction

Comment (kv2019i) [reply sound/soc/sof/intel/hda-pcm.c:298]: @plbossart  seems earlier comment went to /dev/null 

The logic maybe right, but to me this is making this even harder to read. 

The Kconfig setting of ALWAYS_ENABLE_DMI_L1 should have no impact to whether a stream is L1 compatible. These two seems really two separate things, one is a setting of the controller, and another a static property of the stream.

This code now basicly says that if ALWAYS_ENABLE_DMI_L1, mark all streams as L1 compatible even if they really are not.

As we do have DMI_L1_COMPATIBLE flag, I'd find it much easier to understand if we set the stream properties correctly and only override the logic when parsing the flags (i.e. if ALWAYS_ENABLE_DMI_L1 is set, we just ignore the stream properties and enable DMI-L1 anyway). But flags would be correct still.

Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:298]: I have the opposite understanding

before we ALWAYS enabled DM1 L1, but only for compatible streams. That makes no sense to me. Either we intended to unconditionally enable L1 on ALL streams, or this concept should be removed.

@ranj063 can you please chime in, seems we are confused on the initial concept.

Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:298]: @ranj063 @kv2019i this PR is stuck on this comment, I'd like to start closing on opens so that we can send an RFC to Takashi and Mark.

Comment (ranj063) [reply sound/soc/sof/intel/hda-pcm.c:298]: @plbossart , @harshapriya-n  and I are scheduling experiments with this PR as I type this. Let me get back to you with results and feedback as soon as we have them!

Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:298]: @ranj063 @kv2019i seems we have agreement that those patches are needed, can we try and close on this comment, I don't really get what the objection or counter-proposal might be.

Comment (ranj063) [reply sound/soc/sof/intel/hda-pcm.c:298]: @plbossart I'm good with this change. I think the kconfig does have an impact on whether the stream is DMI compatible or not. For example, if there's no pause, all streams are dmi l1 compatible.

**Comment by plbossart** [comment]
rebased branch to force a new CI check.

File: sound/soc/sof/Kconfig
Comment (lgirdwood) [inline sound/soc/sof/Kconfig:65]: I'm assuming this is Y by default ? If not, we should make this a stronger Y in the text i.e. only select N if you are not running one of the above sound servers and need pause.

File: sound/soc/sof/pcm.c
Comment (lgirdwood) [inline sound/soc/sof/pcm.c:26]: again, should this be an "enable_pause" ?

File: sound/soc/sof/Kconfig
Comment (plbossart) [reply sound/soc/sof/Kconfig:65]: Our CI tests pause_push/pause_release and I'd like to keep it this way.
I'd prefer to keep this opt-in.

Maybe we can reword the text as 

````
 This option disables ALSA pause push/release capabilities for
  SOF drivers. 
 Users or distributions relying on sound servers such as PulseAudio, PipeWire and CRAS should enable 
  this option, since those servers don't use pause_push/pause_release. 
  Others who rely on ALSA-level tests and/or wish to keep support for pause_push/pause_release should keep this option disabled.
  If unsure, select "N"
````

File: sound/soc/sof/intel/hda-pcm.c
Comment (kv2019i) [reply sound/soc/sof/intel/hda-pcm.c:298]: @plbossart sorry for late ack. @ranj063 I get that scenario (no pause -> stream compatibility is changed), but if I set "hda_always_enable_dmi_l1=true" via modprobe.d, this function will now set all streams as L1 compatible, even if they are not (me setting a module parameter does not change the stream properties). Effectively we are ignoring the compatibility check and just enabling DMI-L1 anyways, but that's where the SOF_HDA_STREAM_DMI_L1_COMPATIBLE flag seems misleading. But oh well, this is internal implementation and after the kconfig options were renamed in the series, this is easier to follow now, so I'm ok with this.

File: sound/soc/sof/Kconfig
Comment (kv2019i) [reply sound/soc/sof/Kconfig:65]: Sorry for the late comment again, but looking at this, this seems very hard for the integrators.. why such an option? For rewinds, the link to power savings is more understandable, but pause/release is just weird. Would it be easier if we'd just silently disable pause for affected Intel platforms and have some HDA specific option to force enabling of pause (with a fat warning that enabling this option might cause increased power usage). Then we could add that option to sof default kconfig and be able to continue to run the sof-test cases for pause push/release, and rest of the world would get optimal power consumption out-of-the-box.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Sorry, late comments -- I'm ok with the always-dmi-l1 logic, but I'm not sure about the top-level disable-pause kconfig option.

File: sound/soc/sof/Kconfig
Comment (plbossart) [reply sound/soc/sof/Kconfig:65]: This is one of those Brexit or endianness questions....
My approach was: 
when you use make olddefconfig, you get the *same* behavior as before. I am in favor of opt-in options when the behavior changes, something I learned the hard way when I broke audio on Linus' laptop when the HDaudio legacy driver was no longer selected and the other driver didn't do so well.

I don't mind changing this if ALSA maintainers view this as confusing, so we can try and revert to default if there is consensus.

Comment (kv2019i) [reply sound/soc/sof/Kconfig:65]: @plbossart Ok, fair enough -- it's true just dropping pause would change behaviour.

**Review by kv2019i — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]
Thank you @plbossart. I ran the multi pipeline tests with this PR and it passed with flying colors on my TGL soundwire device

**Comment by plbossart** [comment]
Thanks @ranj063 let's merge and do some more tests.

**Pull request merged**
